Assignment4/Diego_Ortega_Martinez_707214/Task07.py

```diff
@@ -0,0 +1,217 @@
+from validation import Report
+#%% md
+# **Task 07: Querying RDF(s)**
+#%%
+import urllib.request
+url = 'https://raw.githubusercontent.com/FacultadInformatica-LinkedData/Curso2025-2026/refs/heads/master/Assignment4/course_materials/python/validation.py'
+urllib.request.urlretrieve(url, 'validation.py')
+github_storage = "https://raw.githubusercontent.com/FacultadInformatica-LinkedData/Curso2025-2026/master/Assignment4/course_materials"
+#%%
+from validation import Report
+#%% md
+# First let's read the RDF file
+#%%
+from rdflib import Graph, Namespace
+
+# Do not change the name of the variables
+g = Graph()
+g.namespace_manager.bind('ns', Namespace("http://somewhere#"), override=False)
+g.parse(github_storage+"/rdf/data06.ttl", format="TTL")
+report = Report()
+#%% md
+# **TASK 7.1a: For all classes, list each classURI. If the class belogs to another class, then list its superclass.**
+# **Do the exercise in RDFLib returning a list of Tuples: (class, superclass) called "result". If a class does not have a super class, then return None as the superclass**
+#%%
+from rdflib import Namespace, URIRef
+from rdflib.namespace import RDF, RDFS
+
+ontology = Namespace("http://oeg.fi.upm.es/def/people#")
+ONT = str(ontology)
+
+def in_ontology(u):
+    return isinstance(u, URIRef) and str(u).startswith(ONT)
+
+
+
+classes = set()
+
+for c in g.subjects(RDF.type, RDFS.Class):
+    if in_ontology(c):
+        classes.add(c)
+
+for c in g.subjects(RDFS.subClassOf, None):
+    if in_ontology(c):
+        classes.add(c)
+
+for c in g.objects(None, RDFS.subClassOf):
+    if in_ontology(c):
+        classes.add(c)
+
+
+result = []
+for c in sorted(classes, key=lambda u: str(u)):
+
+    supers = [s for s in g.objects(c, RDFS.subClassOf) if in_ontology(s)]
+    sup = supers[0] if supers else None
+    result.append((c, sup))
+
+for r in result:
+    print(r)
+
+#%%
+## Validation: Do not remove
+report.validate_07_1a(result)
+#%% md
+# **TASK 7.1b: Repeat the same exercise in SPARQL, returning the variables ?c (class) and ?sc (superclass)**
+#%%
+query = """
+PREFIX rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX ontology: <http://oeg.fi.upm.es/def/people#>
+
+SELECT ?c (SAMPLE(?sc) AS ?sc) WHERE {
+  {
+    ?c a rdfs:Class .
+  } UNION {
+    ?c rdfs:subClassOf ?x .
+  } UNION {
+    ?x rdfs:subClassOf ?c .
+  }
+
+  FILTER(STRSTARTS(STR(?c), STR(ontology:)))
+
+  OPTIONAL {
+    ?c rdfs:subClassOf ?sc .
+    FILTER(STRSTARTS(STR(?sc), STR(ontology:)))
+  }
+}
+GROUP BY ?c
+ORDER BY ?c
```

Comment:
no need to group, or order, or have filters.
Repeat the query using rdfs:subclassOf*